streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -439,7 +440,20 @@ public Subscription subscription(final Set<String> topics) {
                                     } else {
                                         numPartitionsCandidate = metadata.partitionCountForTopic(sourceTopicName);
                                         if (numPartitionsCandidate == null) {
-                                            repartitionTopicMetadata.get(topicName).numPartitions = NOT_AVAILABLE;
+                                          assignment = new HashMap<>();
```

Comment:
This is a meta comment: I'd suggest having a separate check at the very beginning of `assign()`, after `Step Zero`, that for each entry value in `topicGroups = taskManager.builder().topicGroups()`, if each of its `TopicsInfo#sourceTopics` are either in `TopicsInfo#repartitionSourceTopics` or can be found in `metadata`. If the check fails we immediately falls back into the error case of 1) log an ERROR, and 2) set dummy assignment to all the clients with error code. 

Then in line 419 here we do not need this `do-while` loop, instead we should follow the sub-topology id ordering to assign num.partitions for repartition topics, and if it cannot be decided we will throw an runtime exception since it is not expected any more. In addition we can remove `NOT_AVAILABLE` as well.

Reply:
Still need to figure out how to follow the sub-topology id ordering.
For now, I keep the do-while loop.

Reply:
@tedyu @vvcephei one way to break the while loop, is to rely on the 

```
Map<Integer, InternalTopologyBuilder.TopicsInfo> topicGroups
```

Note the `Integer` key is indeed the sub-topology id here, and since we sort the sub-topology by their ids, starting with `1`, the first sub-topology 1 should have no internal topics as its source topic. 

So we can start by this key ordering, to first determine any of the repartition topic's num.partitions as their sink topics of sub-topology 1, and then based on them as for the source topics of sub-topology 2, we can determine sub-topology 2's sink repartition topic's numb.partitions, and so on.